TijanaSonjic/assignment3/wordSearch.cpp

```diff
@@ -0,0 +1,160 @@
+#include "iostream"
+#include "set"
+#include "vector"
+#include <algorithm>
+#include <string>
+#include "test_libraryr.h"
+
+using namespace std;
+
+
+
+set<string> dictionary = { "apple", "spring", "sun", "sea", "dancer" };
+
+typedef struct gridCell {
+	char letter;
+	bool visited;
+} GridCell;
+
+typedef struct delta {
+	int dx;
+	int dy;
+} Delta;
+
+GridCell grid[][3] = {
+						{ {'s', false}, {'p', false}, {'g', false} },
+						{ {'t', false}, {'r', false}, {'n', false} },
+						{ {'s', false}, {'u', false}, {'i', false} }
+					};
+
+// Directions for adjacent cells
+Delta difference[8] = { {-1,-1},{-1,0},{-1,1},{0,-1},{0,1},{1,-1},{1,0},{1,1} };
+
+// Discard words that have letters which are not in the grid.
+//
+set<string> reduceDictionary(set<string> dictionary, set<char> gridLetters) {
+	set<string> result;
+
+	for (set<string>::iterator it = dictionary.begin(); it != dictionary.end(); ++it) {
+		int cnt = 0;
+		string word = *it;
+		transform(word.begin(), word.end(), word.begin(), ::tolower);
+
+		for (int i = 0; i < word.length(); i++) {
+			if (gridLetters.find(word[i]) == gridLetters.end())
+				break;
+			cnt++;
+		}
+
+		if (word.length() == cnt)
+			result.insert(word);
+	}
+
+	return result;
+}
+
+// This allows the search function to abort search paths which can never result in words.
+//
+set<string> generateWordsPrefixes(set<string> dictionary) {
+	set<string> result;
+
+	for (set<string>::iterator it = dictionary.begin(); it != dictionary.end(); ++it) {
+		string word = *it;
+		for (int i = 0; i <= word.length(); i++) {
+			result.insert(word.substr(0, i));
+		}
+	}
+
+	return result;
+}
+
+set<char> makeGridSet(GridCell grid[][3]) {
+	set<char> result;
+	for (int i = 0; i < 3; i++)
+		for (int j = 0; j < 3; j++) {
+			result.insert(grid[i][j].letter);
+		}
+	return result;
+}
+
+bool isPrefix(set<string> prefixes, string word) {
+	return prefixes.find(word) != prefixes.end();
+}
+
+bool isWord(set<string> dictionary, string word) {
+	return dictionary.find(word) != dictionary.end();
+}
+
+bool check(int i, int j) {
+	return i >= 0 && i < 3 && j >= 0 && j < 3 && !grid[i][j].visited;
+}
+
+void findWords(set<string> dictionary, set<string> prefixes, vector<string>& matches, int i, int j, vector<char> generatedWord) {
+	bool validWord = true;
+
+	generatedWord.push_back(grid[i][j].letter);
+	
+	string word(generatedWord.begin(), generatedWord.end());
+
+	if (!isPrefix(prefixes, word)) {
+		return;
+	}
+
+	if (isWord(dictionary, word)) {
+		matches.push_back(word);
+	}
+
+	grid[i][j].visited = true;
+	vector<char> newWord(generatedWord);
+
+	for (int k = 0; k < 8; k++) {
+		bool validToMove = check(i + difference[k].dx, j + difference[k].dy);
```

Comment:
You use both "i,j", and "x,y" coordinates to represent location on the grid. This can be confusing, it is best to stick to one method and keep it.
It is even more confusing, because `i` represents the row, meaning changing of `i` is going up and down; and `j` represents the column, meaning changing `j` is going left and right;
So it means that `j` should correspond with `x`, and `i` should correspond with `y`!
Confused? Me too! ;)

Therefore I recommend changing `Delta` struct fields to be `d_i, d_j` or `diff_i,diff_j`, and not use `x,y` at all.